File: python/understanding_computation/dfa.py
class FaRule:
    def __init__(self, state, character, next_state):
        self.state = state
        self.character = character
        self.next_state = next_state

    def applies_to(self, state, character):
        return self.state == state and self.character == character

    def follow(self):
        return self.next_state

    def __str__(self):
        return "<FaRule: " + str(self.state) + "--" + str(self.character) + "-->" + str(self.next_state)

class DfaRuleBook:
    def __init__(self, rules):
        self.rules = rules

    def next_state(self, state, character):
        return self.rule_for(state, character).follow()

    def rule_for(self, state, character):
        result = list(filter(lambda rule: rule.applies_to(state, character), self.rules))[0]
        return result

class Dfa:
    def __init__(self, current_state, accept_states, rulebook):
        self.current_state = current_state
        self.accept_states = accept_states
        self.rulebook = rulebook

    def accepting(self):
        return self.current_state in self.accept_states

    def read_character(self, character):
        self.current_state = self.rulebook.next_state(self.current_state, character)

    def read_string(self, string):
        for char in string:
            self.read_character(char)


class DfaDesign:
    def __init__(self, start_state, accept_states, rulebook):
        self.start_state = start_state
        self.accept_states = accept_states
        self.rulebook = rulebook

    def to_dfa(self):
        return Dfa(self.start_state, self.accept_states, self.rulebook)

    def accepts(self, string):
        dfa = self.to_dfa()
        dfa.read_string(string)
        return dfa.accepting()

class NfaRulebook:
    def __init__(self, rules):
        self.rules = rules

    def follow_rules_for(self, state, character):
        return list(map(lambda x: x.follow(), self.rules_for(state, character)))

    def rules_for(self, state, character):
        return filter(lambda rule: rule.applies_to(state, character), self.rules)


def main():
    rulebook = NfaRulebook([
        FaRule(1, '1', 1), FaRule(1, 'b', 1), FaRule(1, 'b', 2)
    ])

if __name__ == '__main__':
    main()

File: typescript/src/regex-compiler/dfa.ts
class FARule {
  constructor(public readonly state: number, public readonly character: string, public readonly nextState: number) {}

  public applyTo(state: number, character: string) {
    return this.state === state && this.character === character;
  }
}

class DFARuleBook {
  constructor(public readonly rules: Array<FARule>) {}

  public nextState(state: number, character: string) {
    return this.ruleFor(state, character)?.nextState;
  }

  public ruleFor(state: number, character: string) {
    for (const rule of this.rules) {
      if (rule.applyTo(state, character)) {
        return rule;
      }
    }
  }
}

class DFA {
  constructor(
    public currentState: number,
    public readonly acceptStates: Array<number>,
    public readonly ruleBook: DFARuleBook
  ) {}

  public accepting() {
    return this.acceptStates.includes(this.currentState);
  }

  public readCharacter(character: string) {
    this.currentState = this.ruleBook.nextState(this.currentState, character) || -1;
  }

  public readString(s: string) {
    for (const char of s) {
      this.readCharacter(char);
    }
  }
}

class DFADesign {
  constructor(public startState: number, public acceptStates: Array<number>, public readonly ruleRook: DFARuleBook) {}

  public toDfa() {
    return new DFA(this.startState, this.acceptStates, this.ruleRook);
  }

  public accepts(s: string) {
    const dfa = this.toDfa();
    dfa.readString(s);
    return dfa.accepting();
  }
}

const ruleBook = new DFARuleBook([
  new FARule(1, "a", 2),
  new FARule(1, "b", 1),
  new FARule(2, "a", 2),
  new FARule(2, "b", 3),
  new FARule(3, "a", 3),
  new FARule(3, "b", 3),
]);

class NFARuleBook {
  constructor(public readonly rules: Array<FARule>) {}

  public nextStates(states: Set<number>, character: string) {
    const s = Array.from(states);
    return new Set<number>(s.flatMap((state) => this.followRulesFor(state, character)));
  }

  public followRulesFor(state: number, character: string) {
    return this.rulesFor(state, character).map((rule) => rule.nextState);
  }

  public rulesFor(state: number, character: string) {
    return this.rules.filter((rule) => rule.applyTo(state, character));
  }
}

console.log(ruleBook.nextState(1, "a"));
console.log(ruleBook.nextState(1, "b"));
console.log(ruleBook.nextState(2, "b"));

const dfa = new DFA(1, [3], ruleBook);
console.log(dfa.accepting());
dfa.readString("baaab");
console.log(dfa.accepting());

const dfaDesign = new DFADesign(1, [3], ruleBook);
console.log(dfaDesign.accepts("a"));
console.log(dfaDesign.accepts("baa"));
console.log(dfaDesign.accepts("baba"));

const nfaRuleBook = new NFARuleBook([
  new FARule(1, "a", 1),
  new FARule(1, "b", 1),
  new FARule(1, "b", 2),
  new FARule(2, "a", 3),
  new FARule(2, "b", 3),
  new FARule(3, "a", 4),
  new FARule(3, "b", 4),
]);

console.log(nfaRuleBook.nextStates(new Set([1]), "b"));
console.log(nfaRuleBook.nextStates(new Set([1, 2]), "a"));

class NFA {
  constructor(
    public currentStates: Set<number>,
    public acceptStates: Set<number>,
    public readonly ruleBook: NFARuleBook
  ) {}

  public accepting() {
    const c = Array.from(this.currentStates);
    const a = Array.from(this.acceptStates);
    const intersectionLength = c.filter((it) => a.includes(it)).length;
    return intersectionLength > 0;
  }

  public readCharacter(character: string) {
    this.currentStates = this.ruleBook.nextStates(this.currentStates, character);
  }

  public readString(s: string) {
    for (const char of s) {
      this.readCharacter(char);
    }
  }
}

console.log("NFA -------------------");
console.log(new NFA(new Set([1]), new Set([4]), nfaRuleBook).accepting());
console.log(new NFA(new Set([1, 2, 4]), new Set([4]), nfaRuleBook).accepting());

let nfa = new NFA(new Set([1]), new Set([4]), nfaRuleBook);
console.log(nfa.accepting());
nfa.readCharacter("b");
console.log(nfa.accepting());
nfa.readCharacter("a");
console.log(nfa.accepting());
nfa.readCharacter("b");
console.log(nfa.accepting());
nfa = new NFA(new Set([1]), new Set([4]), nfaRuleBook);
nfa.readString("bbbbb");
console.log(nfa.accepting());

class NFADesign {
  constructor(
    public readonly startState: number,
    public readonly acceptStates: Set<number>,
    public readonly ruleBook: NFARuleBook
  ) {}

  public accepts(s: string) {
    const nfa = this.toNfa();
    nfa.readString(s);
    return nfa.accepting();
  }

  public toNfa() {
    return new NFA(new Set([this.startState]), this.acceptStates, this.ruleBook);
  }
}

console.log("NFADesign ------------------");
const nfaDesign = new NFADesign(1, new Set([4]), nfaRuleBook);
console.log(nfaDesign.accepts("bab"));
console.log(nfaDesign.accepts("bbbbb"));
console.log(nfaDesign.accepts("bbabb"));
